core/src/smartcontracts/wasm.rs

```diff
@@ -330,26 +316,41 @@ impl<'wrld> Runtime<'wrld> {
     ) -> Result<(), Trap> {
         let memory = Self::get_memory(&mut caller)?;
 
-        let instruction = Self::decode_from_memory(&memory, &caller, offset, len)?;
+        let instruction: Instruction = Self::decode_from_memory(&memory, &caller, offset, len)?;
+        debug!(%instruction, "Executing");
 
         let State {
             wsv,
             account_id,
             validator,
+            operation_to_validate,
             ..
         } = caller.data_mut();
 
-        if let Some(validator) = validator {
-            validator
-                .validate_instruction(account_id, &instruction, wsv)
-                .map_err(|error| Trap::new(error.to_string()))?;
+        // TODO: Bad code, need to refactor the whole module
+        let is_permission_validator = operation_to_validate.is_some();
+        if is_permission_validator {
```

Comment:
for me this is super weird. If this smart contract is a permission validator we go and execute the instructions but if it's not a permission validator we go and validate the instruction?

Can you explain?

Reply:
how would a trigger execute and instruction?

Reply:
Yes, it's weird because in order to validate instructions we need to execute them (because one instructions can be dependent from a previous one). That's why validator executes instruction. That's why it's enough to just validate instruction for a smart contract -- instruction will be executed by validation.

And that's why @appetrosyan has suggested to rename it to verifier.

Does all of this make sense for you? Or you have an idea how to improve it?